From 043a1458fce53c1728a4bbbc48d1083fe23532ba Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Tue, 1 Sep 2020 08:51:17 +0800
Subject: [PATCH 71/98] HACK: backend-drm: Treat LINEAR and INVALID modifiers
 as equivalent

Treat DRM_FORMAT_MOD_LINEAR and DRM_FORMAT_MOD_INVALID as equivalent
modifiers to work around Rockchip BSP driver and package expectations.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 clients/simple-dmabuf-egl.c         | 5 +++--
 clients/simple-dmabuf-v4l.c         | 2 +-
 libweston/backend-drm/drm-gbm.c     | 3 ++-
 libweston/backend-drm/fb.c          | 6 ++++--
 libweston/backend-drm/kms.c         | 3 +++
 libweston/linux-dmabuf.c            | 3 +--
 libweston/pixel-formats.c           | 2 +-
 libweston/pixman-renderer.c         | 6 ++++--
 libweston/renderer-gl/gl-renderer.c | 6 +++---
 shared/weston-drm-fourcc.h          | 4 ++++
 10 files changed, 26 insertions(+), 14 deletions(-)

diff --git a/clients/simple-dmabuf-egl.c b/clients/simple-dmabuf-egl.c
index db22a210..435d1f1b 100644
--- a/clients/simple-dmabuf-egl.c
+++ b/clients/simple-dmabuf-egl.c
@@ -265,7 +265,8 @@ create_fbo_for_buffer(struct display *display, struct buffer *buffer)
 	attribs[atti++] = (int) buffer->offsets[plane_idx]; \
 	attribs[atti++] = EGL_DMA_BUF_PLANE ## plane_idx ## _PITCH_EXT; \
 	attribs[atti++] = (int) buffer->strides[plane_idx]; \
-	if (display->egl.has_dma_buf_import_modifiers) { \
+	if (DRM_MOD_VALID(buffer->modifier) && \
+	    display->egl.has_dma_buf_import_modifiers) { \
 		attribs[atti++] = EGL_DMA_BUF_PLANE ## plane_idx ## _MODIFIER_LO_EXT; \
 		attribs[atti++] = buffer->modifier & 0xFFFFFFFF; \
 		attribs[atti++] = EGL_DMA_BUF_PLANE ## plane_idx ## _MODIFIER_HI_EXT; \
@@ -1023,7 +1024,7 @@ dmabuf_modifiers(void *data, struct zwp_linux_dmabuf_v1 *zwp_linux_dmabuf,
 
 	d->format_supported = true;
 
-	if (modifier != DRM_FORMAT_MOD_INVALID) {
+	if (DRM_MOD_VALID(modifier)) {
 		++d->modifiers_count;
 		d->modifiers = realloc(d->modifiers,
 				       d->modifiers_count * sizeof(*d->modifiers));
diff --git a/clients/simple-dmabuf-v4l.c b/clients/simple-dmabuf-v4l.c
index 8ed3a0b5..7a4aa9c3 100644
--- a/clients/simple-dmabuf-v4l.c
+++ b/clients/simple-dmabuf-v4l.c
@@ -912,7 +912,7 @@ dmabuf_modifier(void *data, struct zwp_linux_dmabuf_v1 *zwp_linux_dmabuf,
 	struct display *d = data;
 	uint64_t modifier = u64_from_u32s(modifier_hi, modifier_lo);
 
-	if (format == d->drm_format && modifier == DRM_FORMAT_MOD_LINEAR)
+	if (format == d->drm_format && !DRM_MOD_VALID(modifier))
 		d->requested_format_found = true;
 }
 
diff --git a/libweston/backend-drm/drm-gbm.c b/libweston/backend-drm/drm-gbm.c
index 711d2c12..da219824 100644
--- a/libweston/backend-drm/drm-gbm.c
+++ b/libweston/backend-drm/drm-gbm.c
@@ -283,7 +283,8 @@ create_gbm_surface(struct gbm_device *gbm, struct drm_output *output)
 		return;
 	}
 
-	if (!weston_drm_format_has_modifier(fmt, DRM_FORMAT_MOD_INVALID)) {
+	if (!weston_drm_format_has_modifier(fmt, DRM_FORMAT_MOD_INVALID) &&
+	    !weston_drm_format_has_modifier(fmt, DRM_FORMAT_MOD_LINEAR)) {
 		modifiers = weston_drm_format_get_modifiers(fmt, &num_modifiers);
 		output->gbm_surface =
 			gbm_surface_create_with_modifiers(gbm,
diff --git a/libweston/backend-drm/fb.c b/libweston/backend-drm/fb.c
index 007296fe..c5c4ee95 100644
--- a/libweston/backend-drm/fb.c
+++ b/libweston/backend-drm/fb.c
@@ -243,7 +243,7 @@ drm_fb_addfb(struct drm_device *device, struct drm_fb *fb)
 
 	/* If we have a modifier set, we must only use the WithModifiers
 	 * entrypoint; we cannot import it through legacy ioctls. */
-	if (device->fb_modifiers && fb->modifier != DRM_FORMAT_MOD_INVALID) {
+	if (device->fb_modifiers && DRM_MOD_VALID(fb->modifier)) {
 		/* KMS demands that if a modifier is set, it must be the same
 		 * for all planes. */
 		for (i = 0; i < ARRAY_LENGTH(mods) && fb->handles[i]; i++)
@@ -425,6 +425,7 @@ drm_fb_get_from_dmabuf_attributes(struct dmabuf_attributes *attributes,
 		.modifier = attributes->modifier,
 	};
 
+#if 0
 	/* We should not import to KMS a buffer that has been allocated using no
 	 * modifiers. Usually drivers use linear layouts to allocate with no
 	 * modifiers, but this is not a rule. The driver could use, for
@@ -438,6 +439,7 @@ drm_fb_get_from_dmabuf_attributes(struct dmabuf_attributes *attributes,
 				FAILURE_REASONS_DMABUF_MODIFIER_INVALID;
 		return NULL;
 	}
+#endif
 
 	/* XXX: TODO:
 	 *
@@ -706,7 +708,7 @@ drm_fb_compatible_with_plane(struct drm_fb *fb, struct drm_plane *plane,
 		 * wl_drm is being used for scanout. Mesa is the only user we
 		 * care in this case (even though recent versions are also using
 		 * dmabufs), and it should know better what works or not. */
-		if (fb->modifier == DRM_FORMAT_MOD_INVALID)
+		if (!DRM_MOD_VALID(fb->modifier))
 			return true;
 
 		if (weston_drm_format_has_modifier(fmt, fb->modifier))
diff --git a/libweston/backend-drm/kms.c b/libweston/backend-drm/kms.c
index 64b50d38..56ab8c33 100644
--- a/libweston/backend-drm/kms.c
+++ b/libweston/backend-drm/kms.c
@@ -707,6 +707,9 @@ fallback:
 		ret = weston_drm_format_add_modifier(fmt, DRM_FORMAT_MOD_INVALID);
 		if (ret < 0)
 			return -1;
+		ret = weston_drm_format_add_modifier(fmt, DRM_FORMAT_MOD_LINEAR);
+		if (ret < 0)
+			return -1;
 	}
 	return 0;
 }
diff --git a/libweston/linux-dmabuf.c b/libweston/linux-dmabuf.c
index c460ff58..b2bd0b3f 100644
--- a/libweston/linux-dmabuf.c
+++ b/libweston/linux-dmabuf.c
@@ -1073,8 +1073,7 @@ bind_linux_dmabuf(struct wl_client *client,
 								  fmt->format,
 								  modifier_hi,
 								  modifier_lo);
-			} else if (modifiers[i] == DRM_FORMAT_MOD_LINEAR ||
-				   modifiers[i] == DRM_FORMAT_MOD_INVALID) {
+			} else if (!DRM_MOD_VALID(modifiers[i])) {
 				zwp_linux_dmabuf_v1_send_format(resource,
 								fmt->format);
 			}
diff --git a/libweston/pixel-formats.c b/libweston/pixel-formats.c
index c88da044..91ce8af7 100644
--- a/libweston/pixel-formats.c
+++ b/libweston/pixel-formats.c
@@ -1135,7 +1135,7 @@ pixel_format_get_modifier(uint64_t modifier)
 		return mod_str;
 	}
 
-	if (modifier == DRM_FORMAT_MOD_LINEAR) {
+	if (!DRM_MOD_VALID(modifier)) {
 		str_printf(&mod_str, "%s (0x%llx)", modifier_name,
 			   (unsigned long long) modifier);
 		free(modifier_name);
diff --git a/libweston/pixman-renderer.c b/libweston/pixman-renderer.c
index 4aad0015..a1a20e81 100644
--- a/libweston/pixman-renderer.c
+++ b/libweston/pixman-renderer.c
@@ -748,8 +748,7 @@ pixman_renderer_prepare_dmabuf(struct linux_dmabuf_buffer *dmabuf)
 	int i;
 
 	/* Only support linear or implicit modifiers */
-	if (attributes->modifier != DRM_FORMAT_MOD_INVALID &&
-	    attributes->modifier != DRM_FORMAT_MOD_LINEAR)
+	if (DRM_MOD_VALID(attributes->modifier))
 		return false;
 
 	/* reject all flags we do not recognize or handle */
@@ -1338,6 +1337,9 @@ populate_supported_dmabuf_formats(struct weston_compositor *ec,
 		ret = weston_drm_format_add_modifier(fmt, DRM_FORMAT_MOD_INVALID);
 		if (ret < 0)
 			goto out;
+		ret = weston_drm_format_add_modifier(fmt, DRM_FORMAT_MOD_LINEAR);
+		if (ret < 0)
+			goto out;
 	}
 
 out:
diff --git a/libweston/renderer-gl/gl-renderer.c b/libweston/renderer-gl/gl-renderer.c
index 14a54676..8d3c8959 100644
--- a/libweston/renderer-gl/gl-renderer.c
+++ b/libweston/renderer-gl/gl-renderer.c
@@ -3808,7 +3808,7 @@ import_simple_dmabuf(struct gl_renderer *gr,
 	attribs[atti++] = EGL_IMAGE_PRESERVED_KHR;
 	attribs[atti++] = EGL_TRUE;
 
-	if (attributes->modifier != DRM_FORMAT_MOD_INVALID) {
+	if (DRM_MOD_VALID(attributes->modifier)) {
 		if (!egl_display_has(gr, EXTENSION_EXT_IMAGE_DMA_BUF_IMPORT_MODIFIERS))
 			return NULL;
 		has_modifier = true;
@@ -4283,7 +4283,7 @@ gl_renderer_import_dmabuf(struct weston_compositor *ec,
 	assert(egl_display_has(gr, EXTENSION_EXT_IMAGE_DMA_BUF_IMPORT));
 
 	/* return if EGL doesn't support import modifiers */
-	if (dmabuf->attributes.modifier != DRM_FORMAT_MOD_INVALID)
+	if (DRM_MOD_VALID(dmabuf->attributes.modifier))
 		if (!egl_display_has(gr, EXTENSION_EXT_IMAGE_DMA_BUF_IMPORT_MODIFIERS))
 			return false;
 
@@ -4365,7 +4365,7 @@ populate_supported_dmabuf_formats(struct weston_compositor *ec,
 
 		for (j = 0; j < num_modifiers; j++) {
 			/* Skip MOD_INVALID, as it has already been added. */
-			if (modifiers[j] == DRM_FORMAT_MOD_INVALID)
+			if (!DRM_MOD_VALID(modifiers[j]))
 				continue;
 			ret = weston_drm_format_add_modifier(fmt, modifiers[j]);
 			if (ret < 0) {
diff --git a/shared/weston-drm-fourcc.h b/shared/weston-drm-fourcc.h
index 286886f6..f1edf831 100644
--- a/shared/weston-drm-fourcc.h
+++ b/shared/weston-drm-fourcc.h
@@ -30,6 +30,10 @@
 
 #include <drm_fourcc.h>
 
+/* Check if modifier is neither LINEAR nor INVALID (i.e., a "real" modifier) */
+#define DRM_MOD_VALID(mod) \
+	((mod) != DRM_FORMAT_MOD_LINEAR && (mod) != DRM_FORMAT_MOD_INVALID)
+
 /* The kernel header drm_fourcc.h defines the DRM formats below.  We duplicate
  * some of the definitions here so that building Weston won't require
  * bleeding-edge kernel headers.
-- 
2.39.5

